go-fuzz found panic in SpecificDatumReader mapUnion

```
panic: runtime error: index out of range

goroutine 1 [running]:
gopkg.in/avro%2ev0.sDatumReader.mapUnion(0x11f9800, 0xc42000a2a0, 0x112df60, 0xc420092140, 0x194, 0x11fa880, 0xc42000a340, 0x1146120, 0x1146120, 0x1, ...)
	/gopath/src/gopkg.in/avro.v0/datum_reader.go:354 +0x223
gopkg.in/avro%2ev0.sDatumReader.readValue(0x11f9800, 0xc42000a2a0, 0x112df60, 0xc420092140, 0x194, 0x11fa880, 0xc42000a340, 0x194, 0x0, 0x0, ...)
	/gopath/src/gopkg.in/avro.v0/datum_reader.go:204 +0x577
gopkg.in/avro%2ev0.sDatumReader.findAndSet(0x111faa0, 0xc420092100, 0x16, 0xc420080460, 0x11fa880, 0xc42000a340, 0x11f8300, 0xc42000e0b0)
	/gopath/src/gopkg.in/avro.v0/datum_reader.go:169 +0x12a
gopkg.in/avro%2ev0.sDatumReader.fillRecord(0x11f96e0, 0xc420090070, 0x111faa0, 0xc420092100, 0x16, 0x11fa880, 0xc42000a340, 0xc42003de90, 0x100fa28)
	/gopath/src/gopkg.in/avro.v0/datum_reader.go:403 +0xec
gopkg.in/avro%2ev0.(*SpecificDatumReader).Read(0xc42000e7a0, 0x111faa0, 0xc420092100, 0x11fa880, 0xc42000a340, 0x1066f96, 0x5a3849ec)
	/gopath/src/gopkg.in/avro.v0/datum_reader.go:154 +0x292
gopkg.in/avro.v0/fuzzes/specificreadercomplex.Fuzz(0x12ce000, 0x6, 0x200000, 0xc42003df70)
	/gopath/src/gopkg.in/avro.v0/fuzzes/specificreadercomplex/fuzzer.go:17 +0xed
go-fuzz-dep.Main(0x115b598)
	/goroot/src/go-fuzz-dep/main.go:49 +0xad
main.main()
	/gopath/src/gopkg.in/avro.v0/fuzzes/specificreadercomplex/go.fuzz.main/main.go:10 +0x2d
exit status 2
```
with input:

```
	"\x00\x000000"
```

--- a/datum_reader.go
+++ b/datum_reader.go
@@ -346,13 +346,15 @@ func (reader sDatumReader) mapEnum(field Schema, dec Decoder) (reflect.Value, er
 }
 
 func (reader sDatumReader) mapUnion(field Schema, reflectField reflect.Value, dec Decoder) (reflect.Value, error) {
-	unionType, err := dec.ReadInt()
+	unionIndex, err := dec.ReadInt()
 	if err != nil {
-		return reflect.ValueOf(unionType), err
+		return reflect.ValueOf(unionIndex), err
 	}
-
-	union := field.(*UnionSchema).Types[unionType]
-	return reader.readValue(union, reflectField, dec)
+	types := field.(*UnionSchema).Types
+	if unionIndex < 0 || int(unionIndex) >= len(types) {
+		return reflect.Value{}, fmt.Errorf("Invalid union index %d", unionIndex)
+	}
+	return reader.readValue(types[unionIndex], reflectField, dec)
 }
 
 func (reader sDatumReader) mapFixed(field Schema, dec Decoder) (reflect.Value, error) {

